smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,355 @@
+            )
+
+        # Worker polling period in seconds.
+        self._polling_period = 0.1
+
+        # Fork is not a thread safe method.
+        forkserver_available = "forkserver" in mp.get_all_start_methods()
+        start_method = "forkserver" if forkserver_available else "spawn"
+        mp_ctx = mp.get_context(start_method)
+
+        self.env_constructors = env_constructors
+
+        self.error_queue = mp_ctx.Queue()
+        self.parent_pipes = []
+        self.processes = []
+        for idx, env_constructor in enumerate(self.env_constructors):
+            parent_pipe, child_pipe = mp_ctx.Pipe()
+            process = mp_ctx.Process(
+                target=_worker,
+                name=f"Worker<{type(self).__name__}>-<{idx}>",
+                args=(
+                    idx,
+                    sim_name,
+                    CloudpickleWrapper(env_constructor),
+                    auto_reset,
+                    child_pipe,
+                    self.error_queue,
+                    self._polling_period,
+                ),
+            )
+            self.parent_pipes.append(parent_pipe)
+            self.processes.append(process)
+
+            # Daemonic subprocesses quit when parent process quits. However, daemonic
+            # processes cannot spawn children. Hence, `process.daemon` is set to False.
+            process.daemon = False
+            process.start()
+            child_pipe.close()
+
+        self._state = AsyncState.DEFAULT
+
+        # Wait for all environments to successfully startup
+        _, successes = zip(*[pipe.recv() for pipe in self.parent_pipes])
+        self._raise_if_errors(successes)
+
+        # Get and check observation and action spaces
+        observation_space, action_space = self._get_spaces()
+
+        # TODO: This dummy `observation_space` and `action_space` should be removed after they
+        # are properly specified in SMARTS/smarts/env/hiway_env.py:__init__() function.
+        action_space = gym.spaces.Dict(
+            {
+                agent_id: gym.spaces.Box(
+                    np.array([0, 0, -1]), np.array([+1, +1, +1]), dtype=np.float32
+                )  # throttle, break, steering
+                for agent_id in ["Agent1", "Agent2"]
+            }
+        )
+        observation_space = gym.spaces.Box(
+            low=-1, high=1, shape=(256, 256, 3), dtype=np.float32
+        )
+
+        super(AsyncVectorEnv, self).__init__(
+            num_envs=len(env_constructors),
+            observation_space=observation_space,
+            action_space=action_space,
+        )
+
+        # Seed all the environment
+        self.seed(seed)
+
+    def seed(self, seed: int) -> Tuple[int]:
+        """Sets unique seed for each environment.
+
+        Args:
+            seed (int): Seed number.
+
+        Raises:
+            AlreadyPendingCallError: If `seed` is called while another function call is pending.
+        """
```

Comment:
This should document what is returned as well (especially since this is the key difference between this implementation of `seed()` and the implementation in the parent class).